src/cat/cat.c

```diff
@@ -1,6 +1,325 @@
+#include <ctype.h>
+#include <stdbool.h>
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define NAME "cat (canoutils)"
+#define VERSION "1.0.0"
+#define AUTHOR "Akos Szijgyarto (SzAkos04)"
+
+#define print_version()                                                        \
+  do {                                                                         \
+    printf("%s\nversion: %s\nby: %s\n", NAME, VERSION, AUTHOR);                \
+  } while (0)
+
+#define print_help()                                                           \
+  do {                                                                         \
+    printf("Usage: cat [OPTION]... [FILE]...\n");                              \
+    printf("Concatenate FILE(s) to standard output.\n");                       \
+  } while (0)
+
+#define print_incorrect_args()                                                 \
+  do {                                                                         \
+    printf("incorrect arguments\n");                                           \
+    printf("see `cat --help`\n");                                              \
+  } while (0)
+
+#define BUF_MAX_LEN 4096 // max length of a buffer in bytes
+#define PATH_LEN 256     // max length of a path in bytes
+#define ARGS_MAX 16      // number of the max arguments
+#define ARGS_LEN 32      // max length of the arguments in bytes
+
+bool number_nonblank = false;  // number nonempty output lines, overrides -n
+bool show_ends = false;        // display $ at end of each line
+bool number = false;           // number all output lines
+bool squeeze_blank = false;    // suppress repeated empty output lines
+bool show_tabs = false;        // display TAB characters as ^I
+bool show_nonprinting = false; // use ^ and M- notation, except for LFD and TAB
+
+int cat(int filec, char **paths);
+int print_file(char *buf);
+int print_stdin(void);
+
+int main(int argc, char **argv) {
+  if (argc < 2) {
+    // print stdin
+    if (print_stdin() != 0) {
+      return 1;
+    }
+    return 0;
+  }
+
+  if (strcmp(argv[1], "--version") == 0) {
+    if (argc != 2) {
+      print_incorrect_args();
+      return 1;
+    }
+
+    print_version();
+    return 0;
+  }
+
+  if (strcmp(argv[1], "--help") == 0) {
+    if (argc != 2) {
+      print_incorrect_args();
+      return 1;
+    }
+
+    print_help();
+    return 0;
+  }
+
+  int filec = 0; // file count
+  char **paths = (char **)malloc(sizeof(char) * PATH_LEN * ARGS_MAX);
```

Comment:
Using argv rotation, it should be achievable without memory allocation

Reply:
Isn't this a better and easier alternative for argv rotation:
`  char *paths[ARGS_MAX];`

Reply:
Sorry, instead of argv rotation

Reply:
> Isn't this a better and easier alternative for argv rotation: ` char *paths[ARGS_MAX];`

Cat can receive as many arguments as any regular comment, which for a linux system is about ~131k

Reply:
And if I define ARGS_MAX around that number? (from what I have found 419430)

Reply:
That would take insane amount of space on the stack!

Reply:
Made it take up `(argc - 1) * sizeof(char *)` bytes at the start, and then reallocated it later to the actual size of the array. Now it can hold up to 4096 strings, and doesn't take up a lot of the heap at the beginning either.